### PR TITLE
Names with locations

### PR DESCRIPTION
**Pull request by fferreira**

### CODE REVIEW

File: lib/codegen.ml [hunk 1]
```diff
@@ -115,11 +115,13 @@ let find_all_payloads g =
 let find_all_roles g =
   let f (_, a, _) acc =
     match a with
-    | SendA (r, _) | RecvA (r, _) -> S.add acc r
+    | SendA (r, _) | RecvA (r, _) -> S.add acc (Name.user r)
     | _ -> failwith "Impossible"
   in
   G.fold_edges_e f g (S.empty (module String)) |> S.to_list
 
+(* assert false *)
```
Comment (fangyi-zhou) [inline lib/codegen.ml]:
```suggestion

```

File: lib/lib.mli [hunk 2]
```diff
@@ -20,15 +20,16 @@ val validate_exn : Syntax.scr_module -> verbose:bool -> unit
 
 (** {1 Other operations} *)
 
-val enumerate : Syntax.scr_module -> (string * string) list
+val enumerate : Syntax.scr_module -> (Syntax.name * Syntax.name) list
```
Comment (fangyi-zhou) [inline lib/lib.mli]: Name.t?

File: lib/loc.ml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+let ghost_loc = (Lexing.dummy_pos, Lexing.dummy_pos)
+
+let sexp_of_source_loc _ = Sexp.Atom "<opaque>"
+
+let render_pos pos =
+  sprintf "%d:%d" pos.Lexing.pos_lnum
+    (pos.Lexing.pos_cnum - pos.Lexing.pos_bol + 1)
+
+let render_pos_interval (startp, endp) : string =
```
Comment (fangyi-zhou) [inline lib/loc.ml]: show_source_loc?

File: lib/loc.ml [hunk 4]
```diff
@@ -0,0 +1,29 @@
+open! Base
+open Printf
+
+type source_loc = Lexing.position * Lexing.position
+
+let ghost_loc = (Lexing.dummy_pos, Lexing.dummy_pos)
+
+let sexp_of_source_loc _ = Sexp.Atom "<opaque>"
+
+let render_pos pos =
```
Comment (fangyi-zhou) [inline lib/loc.ml]: show_position?

File: lib/ltype.ml [hunk 5]
```diff
@@ -45,52 +49,59 @@ let show =
...
+        (* Choice is a set of receive *)
+        merge_recv r1 (lty2 :: ltys1)
+    | RecvL (_, r2, _), ChoiceL (r1, ltys2) when Name.equal r1 r2 ->
+        merge_recv r1 (lty1 :: ltys2)
+    | ChoiceL (r1, ltys1), ChoiceL (r2, ltys2)
+      when Name.equal r1 r2 && not (Name.equal r1 projected_role) ->
+        merge_recv r1 (ltys1 @ ltys2)
+    | _ -> if equal lty1 lty2 then lty1 else fail ()
+  with Unmergable (l1, l2) ->
+    let s1 = Sexp.to_string (sexp_of_t l1) in
```
Comment (fangyi-zhou) [inline lib/ltype.ml]: show

File: lib/ltype.ml [hunk 6]
```diff
@@ -45,52 +49,59 @@ let show =
...
+    let s2 = Sexp.to_string (sexp_of_t l2) in
+    let error = s1 ^ " " ^ s2 in
+    Err.UserError (Err.UnableToMerge error) |> raise
```
Comment (fangyi-zhou) [inline lib/ltype.ml]: uerr

File: lib/parser.mly [hunk 7]
```diff
@@ -44,12 +44,15 @@
 
 (* ---------------------------------------- *)
 %start <Syntax.scr_module> scr_module
-%{ open Syntax
+%{ module S = Syntax
```
Comment (fangyi-zhou) [inline lib/parser.mly]: Why don't you `open S`

Comment (fferreira) [reply lib/parser.mly]: MMMM, I don't have a good answer now, but while I was doing it, it helped separate Syntax.name from the parser name.

Comment (fangyi-zhou) [reply lib/parser.mly]: I see, but now it's ok to open S?

Comment (fferreira) [reply lib/parser.mly]: I ended up opening Syntax

File: lib/err.ml [hunk 8]
```diff
@@ -8,8 +9,9 @@ type user_error =
   | RedefinedRecursionName of string * source_loc * source_loc
   | Uncategorised of string
   | InvalidCommandLineParam of string
-  | UnboundRole of string * source_loc
-  | ReflexiveMessage of string * source_loc
+  | UnboundRole of name * source_loc
```
Comment (fangyi-zhou) [inline lib/err.ml]: Do we need this `source_loc`?

File: lib/gtype.ml [hunk 9]
```diff
@@ -40,13 +43,13 @@ let show =
 let of_protocol global_protocol =
   let {value= {name; roles; interactions; _}; _} = global_protocol in
   let has_global_recursion = ref false in
-  let global_recursion_name = "__" ^ name in
+  let global_recursion_name = Name.rename name ("__" ^ Name.user name) in
```
Comment (fangyi-zhou) [inline lib/gtype.ml]: This will get removed in #46

File: lib/gtype.mli [hunk 10]
```diff
@@ -1,3 +1,5 @@
+open Loc
```
Comment (fangyi-zhou) [inline lib/gtype.mli]: This `loc` is for `located` in L21?

File: lib/loc.ml (hunk 3 above)
Comment (fferreira) [reply lib/loc.ml]: renamed it

File: lib/loc.ml (hunk 4 above)
Comment (fferreira) [reply lib/loc.ml]: renamed it

File: lib/err.ml (hunk 8 above)
Comment (fferreira) [reply lib/err.ml]: No

File: lib/gtype.mli (hunk 10 above)
Comment (fferreira) [reply lib/gtype.mli]: yes, removed

File: lib/gtype.ml (hunk 9 above)
Comment (fferreira) [reply lib/gtype.ml]: ok

**Pull request merged**